acme/api/order.go

```diff
@@ -0,0 +1,186 @@
+package api
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/go-chi/chi"
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/acme"
+	"github.com/smallstep/certificates/api"
+	"github.com/smallstep/certificates/logging"
+)
+
+// NewOrderRequest represents the body for a NewOrder request.
+type NewOrderRequest struct {
+	Identifiers []acme.Identifier `json:"identifiers"`
+	NotBefore   time.Time         `json:"notBefore"`
+	NotAfter    time.Time         `json:"notAfter"`
+}
+
+// Validate validates a new-order request body.
+func (n *NewOrderRequest) Validate() error {
+	if len(n.Identifiers) == 0 {
+		return acme.MalformedErr(errors.Errorf("identifiers list cannot be empty"))
+	}
+	for _, id := range n.Identifiers {
+		if id.Type != "dns" {
+			return acme.MalformedErr(errors.Errorf("identifier type unsupported: %s", id.Type))
+		}
+	}
+	now := time.Now().UTC()
+	if n.NotBefore.IsZero() {
+		n.NotBefore = now
+	}
+	if n.NotAfter.IsZero() {
+		n.NotAfter = now.Add(time.Hour * 24)
+	}
+	if now.After(n.NotAfter) {
+		return acme.MalformedErr(errors.Errorf("NotAfter is already in the past: %s",
+			n.NotAfter))
+	}
+	if n.NotBefore.After(n.NotAfter) {
+		return acme.MalformedErr(errors.Errorf("NotAfter is before NotBefore "+
+			"- NotBefore: %s, NotAfter: %s", n.NotBefore, n.NotAfter))
+	}
+	return nil
+}
+
+// FinalizeRequest captures the body for a Finalize order request.
+type FinalizeRequest struct {
+	CSR string
+	csr *x509.CertificateRequest
+}
+
+// Validate validates a finalize request body.
+func (f *FinalizeRequest) Validate() error {
+	var err error
+	csrBytes, err := base64.RawURLEncoding.DecodeString(f.CSR)
+	if err != nil {
+		return acme.MalformedErr(errors.Wrap(err, "error base64url decoding csr"))
+	}
+	f.csr, err = x509.ParseCertificateRequest(csrBytes)
```

Comment:
are we calling `CheckSignature()` at some point?

Reply:
we are not. because we forgot.